File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/utils/SemVerUtils.java
package io.github.sanyavertolet.edukate.backend.utils;

import reactor.util.function.Tuple3;
import reactor.util.function.Tuples;

public final class SemVerUtils {
    public static String majorFieldName = "majorId";
    public static String minorFieldName = "minorId";
    public static String patchFieldName = "patchId";

    private SemVerUtils() { }

    public static Tuple3<Integer, Integer, Integer> parse(String version) {
        String[] segments = version.split("\\.");
        if (segments.length != 3) {
            throw new IllegalArgumentException("Invalid version format: " + version);
        }
        return Tuples.of(Integer.parseInt(segments[0]), Integer.parseInt(segments[1]), Integer.parseInt(segments[2]));
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/utils/SemVerUtils.kt
package io.github.sanyavertolet.edukate.backend.utils

import org.springframework.data.domain.Sort

object SemVerUtils {
    const val MAJOR_FIELD_NAME = "majorId"
    const val MINOR_FIELD_NAME = "minorId"
    const val PATCH_FIELD_NAME = "patchId"

    private const val SEMVER_SEGMENT_COUNT = 3

    fun semVerSort(
        majorField: String = MAJOR_FIELD_NAME,
        minorField: String = MINOR_FIELD_NAME,
        patchField: String = PATCH_FIELD_NAME,
    ): Sort =
        Sort.by(Sort.Order.asc(majorField), Sort.Order.asc(minorField), Sort.Order.asc(patchField), Sort.Order.asc("_id"))

    fun parse(version: String): Triple<Int, Int, Int> {
        val segments = version.split(".")
        require(segments.size == SEMVER_SEGMENT_COUNT) { "Invalid version format: $version" }
        return Triple(segments[0].toInt(), segments[1].toInt(), segments[2].toInt())
    }
}
